plugins/postgres/dbt/adapters/postgres/connections.py

```diff
@@ -131,23 +148,23 @@ def open(cls, connection):
                 connect_timeout=credentials.connect_timeout,
                 **kwargs,
             )
-
             if credentials.role:
                 handle.cursor().execute("set role {}".format(credentials.role))
-
-            connection.handle = handle
-            connection.state = "open"
-        except psycopg2.Error as e:
-            logger.debug(
-                "Got an error when attempting to open a postgres " "connection: '{}'".format(e)
-            )
-
-            connection.handle = None
-            connection.state = "fail"
-
-            raise dbt.exceptions.FailedToConnectException(str(e))
-
-        return connection
+            return handle
+
+        retryable_exceptions = [
+            # OperationalError is subclassed by all psycopg2 Connection Exceptions
+            # and it's raised by generic connection errors without an error code.
+            psycopg2.errors.OperationalError,
+        ]
```

Comment:
I think you're right that all the exceptions we want to retry are subclasses of `OperationalError`. However, this set includes invalid passwords and the disk being full. This is a smaller subset of exceptions I think we should retry on connection:

ConnectionException
SqlclientUnableToEstablishSqlconnection
ConnectionDoesNotExist
SqlserverRejectedEstablishmentOfSqlconnection
ConnectionFailure
TooManyConnections

Unfortunately the psychopg2 docs don't have much info on these exceptions nor do the postgres docs. I had to go to postgres source code to figure out what these exceptions do and of course, I could be wrong.

Reply:
The problem is that psycopg2 raises `OperationalError`s on generic timeouts (i.e. when waiting for a response from the database upon attempting to connect exceeds `connect_timeout`). The exception message looks like:

This is because specific subclasses are implemented only for errors that have an associated SQLSTATE, or messages emitted by the database server. But as we haven't connected to the database yet, there is no message from the database server!

So we need to catch `OperationalError`. And if we are going to catch `OperationalError` we might as well exclude all other subclasses as they are already caught too.

Reply:
There is an issue on the `psycopg2` GitHub about `OperationalError` being too generic: https://github.com/psycopg/psycopg2/issues/682

But unfortunately it only addressed the creation of individual subclasses for errors that have an associated PGCODE. Errors without a PGCODE (like generic timeouts) have PGCODE set to None, and it doesn't look it will get a subclass:

OP:
> In some cases, particularly around authentication, pgcode is None and so there is no alternative (that I have found) to parsing the error message.

Maintainer:
> No, we don't have plan to add other information other than what the server or the postgres client library provides us, only to make easier to access them.
> I feel your pain though. You should ask the postgres developer to add such a code. Note that those messages are susceptible of localization, apart from being more likely than a code to change from one version to the other.

Reply:
In order to prove this is how timeouts work, I did the following:

1. I found a tool that can simulate a slow TCP proxy: https://github.com/Shopify/toxiproxy. Installed and ran a server.
2. I setup a local connection to a Redshift cluster: `localhost:5439`. Connection is the same I use for development, and works without issues.
3. I setup a proxy that redirects traffic to the local Redshift cluster using the toxiproxy-cli: `toxiproxy-cli create --listen localhost:65432 --upstream localhost:5439 redshift-proxy`
4. I ran the following Python Program (credentials omitted from dsn). Notice that `connect_timeout=1`:

```python
import psycopg2

try:
    conn = psycopg2.connect("host=localhost port=65432 connect_timeout=1")
except Exception as e:
    saved = e
```
Before adding latency this produces a proper connection:

```python
In [3]: conn
Out[3]: <connection object at 0x7f6349895a80; dsn: 'host=localhost port=65432 connect_timeout=1', closed: 0>
```

6. I added some latency: `toxiproxy-cli toxic add -t latency -a latency=10000 redshift-proxy`
7. I re-ran the same Python program as in 4:

```python
In [3]: conn
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 conn

NameError: name 'conn' is not defined

In [4]: str(saved)
Out[4]: 'connection to server at "localhost" (::1), port 65432 failed: Connection refused\n\tIs the server running on that host and accepting TCP/IP connections?\nconnection to server at "localhost" (127.0.0.1), port 65432 failed: timeout expired\n'

In [5]: assert saved.pgcode is None

In [6]: assert saved.pgerror is None

In [7]: saved.__class__
Out[7]: psycopg2.OperationalError
```

This time, due to the added latency being higher than 1, connecting fails. `psycopg2` raises a `psycopg2.OperationalError` with a `timeout expired` error message. This exception object *doesn't have a pgcode*.

Reply:
Well that's rather unfortunate. `OperationalError` it is then. I would add a comment with the link to that bug that's blocking us from being more specific in our retries.

Instead of a `List[Type[Exception]]` we could take a `Callable[[Type[Exception]], bool]` but that might be just be confusing overkill for the interface. I'm fine for the Postgres implementation just passing `[OperationalError]`.